Periodically check active AWX jobs status

zgalor opened the pull request:
This PR is split into 2 commits:

1. Adding job resources to awx pkg.
2. Adding a worker that periodically checks active jobs status.

jhernand commented on pkg/awx/job.go:
```diff
@@ -0,0 +1,32 @@
+	return j.id
+}
+
+func (j *Job) Status() string {
```
Are the status known strings? If they are then I think it would be useful to have constants for them, so that users know what to expect:

```go
type JobStatus string

const (
  JobStatusRunning JobStatus = "running"
  JobStatusFailed JobStatus = "failed"
  ...
)
```
)

jhernand commented on pkg/awx/job_resource.go:
```diff
@@ -0,0 +1,68 @@
+	jobData *Job
+}
+
+func (r *JobGetResponse) JobData() *Job {
```
Can the name of this method be just `Job`?

jhernand commented on pkg/awx/internal/data/job.go:
```diff
@@ -0,0 +1,28 @@
+limitations under the License.
+*/
+
+// This file contains the data structures used for sending and receiving job templates.
```
This ^ comment isn't accurate, should say "jobs", not "job templates".

jhernand commented on manifests/test-alert.json:
```diff
@@ -0,0 +1,32 @@
+{
```
Are you aware of the `*-alert.json` files inside the `examples` directory? If you want to add this new one please put it into the same `examples` directory (you may need to rebase, the `examples` directory was added recently).

jhernand commented on cmd/autoheal/awx_job.go:
```diff
@@ -102,5 +102,29 @@ func (h *Healer) launchAWXJob(connection *awx.Connection, template *awx.JobTempl
 		templateName,
 	)
+
+	glog.Infof(
```
I think that the job id can be included in the previous message:

```
Request to launch AWX job from template 'mytemplate' has been sent, job identifier is '123'
```

jhernand commented on cmd/autoheal/awx_job.go:
```diff
@@ -102,5 +102,29 @@ func (h *Healer) launchAWXJob(connection *awx.Connection, template *awx.JobTempl
+		response.Job,
+	)
+
+	h.watchJobStatus(connection, response.Job)
```
If I understand correctly we are checking the job status just once. I assume that you will evolve that in future versions of the pull request. We will probably need to remember (in memory) the jobs that we created periodically check their status. A possible way to do that is to create another work queue, post to it the jobs that need checking, remove the jobs that have been completed and re-post the ones that aren't completed.

nimrodshn commented on pkg/awx/jobs_resource.go:
```diff
@@ -0,0 +1,85 @@
+	response.previous = output.Previous
+	response.next = output.Next
+	response.results = make([]*Job, len(output.Results))
+	for i := 0; i < len(output.Results); i++ {
```
Just a nitpick - maybe use the `for.. range` syntax?[1] It's a little prettier IMHO.
```
for _, response := range response.results {
   ...
}
```

[1] https://gobyexample.com/range

zgalor commented on cmd/autoheal/healer.go:
```diff
@@ -132,6 +137,7 @@ func (h *Healer) Run(stopCh <-chan struct{}) error {
 	// Start the workers:
 	go wait.Until(h.runRulesWorker, time.Second, stopCh)
 	go wait.Until(h.runAlertsWorker, time.Second, stopCh)
+	go wait.Until(h.RunActiveJobsWorker, time.Minute, stopCh)
```
TODO: make this configurable.
This is 1 minute for debug purposes. default will be 5 minutes

zgalor commented:
@jhernand @nimrodshn @moolitayer 

I rebased and refactored. I think I addressed all of your comments.
This is broken into 2 commits as @jhernand requested.

Still to do:
- make the checking interval configurable (Not sure we want to allow the user to change this)
- remove some of the logs (or condition them with level)
- Add tests? I am actually not sure if and how to unit test this while I depend on awx connection...

Future optimization:
Check if we can get a bulk of active jobs in a single HTTP get

Your reviews are welcome

nimrodshn commented:
If this PR is not ready please add [WIP] to title.

zgalor commented:
I added configuration parameter, rebased, and removed the WIP.
@jhernand @nimrodshn @moolitayer, please review

jhernand commented on cmd/autoheal/awx_job.go:
```diff
@@ -100,16 +100,69 @@ func (h *Healer) launchAWXJob(
 	return nil
 }
+
+func (h *Healer) checkAWXJobStatus(jobID int) (bool, error) {
```
In the rest of the code of the project when a function returns multiple values we name them:

```go
func (h *Healer) checkAWXJobStatus(jobID int) (finished bool, err error) {
   ...
}
```

Can we use the same idiom here?

jhernand commented on cmd/autoheal/awx_job.go:
```diff
@@ -100,16 +100,69 @@ func (h *Healer) launchAWXJob(
+		Insecure(awxInsecure).
+		Build()
+	if err != nil {
+		return false, err
```
If you use the idiom that I suggested in the previous comments, then this could be just `return`.

jhernand commented on cmd/autoheal/active_jobs_worker.go:
```diff
@@ -0,0 +1,51 @@
+	"k8s.io/apimachinery/pkg/util/runtime"
+)
+
+func (h *Healer) RunActiveJobsWorker() {
```
Can this method be private?

jhernand commented on README.md:
```diff
@@ -104,6 +104,11 @@ should always be set to `false` (the default) in production environments.
 The `project` parameter is the name of the AWX project that contains the job
 templates that will be used to run the playbooks.
 
+The autoheal service performs a periodic job status check against AWX server,
```
In the rest of the documentation we write _auto-heal_, with a dash, can you do the same here?

jhernand commented on pkg/config/loader.go:
```diff
@@ -241,6 +252,17 @@ func (l *Loader) mergeAWXCredentials(reference *core.SecretReference) error {
 	return nil
 }
 
+func (l *Loader) mergeAwxJobStatusCheckInterval(intervalStr string) error {
```
I think that this code could be included in the caller, instead of having a new function, as it is too small. Not really important, feel free to ignore.

zgalor commented:
@jhernand I applied all the comments

nimrodshn commented:
@zgalor Looks good :+1: 

@jhernand I would really like for tests to be added to this PR, on the other hand I don't wish to block it and @zgalor says it works for him - If you think it's OK to merge it without tests than we (me and @zgalor) can work together to add tests to this code after it get's merged.

nimrodshn commented:
@jhernand I'm commited for this cause already :smiley:

zgalor commented:
@nimrodshn @jhernand I tested this locally but I'm not sure how we can add automated tests without creating an AWX mock

jhernand commented:
OK, I am merging.
